service/geopmdpy/varrun.py

```diff
@@ -700,11 +661,98 @@ def _load_session_file(self, sess_path):
             if batch_pid is not None and \
                self._is_pid_valid(batch_pid, file_time) == False:
                 sess.pop('batch_server')
+            self._sessions[client_pid] = dict(sess)
+            self._update_session_file(client_pid)
         else:
-            # Invalid session, remove batch server
-            sess['client_pid'] = self._INVALID_PID
-            if batch_pid is not None:
-                sess.pop('batch_server')
+           os.rename(sess_path, renamed_path)
 
-        self._sessions[client_pid] = dict(sess)
-        self._update_session_file(client_pid)
+
+class WriteLock(object):
+    def __init__(self, var_path=GEOPM_SERVICE_VAR_PATH):
+        self._VAR_PATH = var_path
+        self._LOCK_PATH = os.path.join(self._VAR_PATH, "CONTROL_LOCK")
+        self._fid = None
+
+    def __enter__(self):
+        old_mask = os.umask(0o077)
+        try:
+            trial_count = 0
+            while self._fid is None and trial_count < 2:
```

Comment:
Why try twice?  Why not more/less?
Is a sleep necessary between trials?

Reply:
We will try twice, the first time may be foiled by a bad file, if the second attempt is also foiled, then we want to give up.  We want to try more than once to avoid a bad file causing the lock operation to fail.

We don't need to sleep because the is_secure_file() call will rename it to an invalid name.